engine/src/main/java/org/terasology/engine/world/generation/WorldBuilder.java

```diff
@@ -200,92 +209,68 @@ public World build() {
-                    } else {
-                        facetCalculationInProgress.remove(facet);
-                        return;
-                    }
-                }
+                providedBy.put(facet, provider);
+                addRequirements(facet, provider, scalable, orderedProviders);
                 orderedProviders.add(provider);
                 producer = provider;
             }
         }
 
-        if (producer == null) {
-            if (!scalable) {
-                logger.warn("No facet provider found that produces {}", facet);
+        for (FacetProvider provider : providersList) {
+            if (updatesFacet(provider, facet) && (!scalable || provider instanceof ScalableFacetProvider)) {
+                if (updatePriority(provider, facet) > minPriority) {
+                    providedBy.put(facet, provider);
+                    addRequirements(facet, provider, scalable, orderedProviders);
+                    orderedProviders.add(provider);
+                }
             }
-            facetCalculationInProgress.remove(facet);
+        }
+    }
+
+    private void addRequirements(Class<? extends WorldFacet> providedFacet, FacetProvider provider, boolean scalable,
+                                 Set<FacetProvider> orderedProviders) {
+        if (orderedProviders.contains(provider)) {
             return;
         }
 
-        // then add all @Updates facet providers
-        providersList.stream().filter(provider -> updatesFacet(provider, facet)
-                && (!scalable || provider instanceof ScalableFacetProvider)).forEach(provider -> {
-            Set<FacetProvider> localOrderedProviders = Sets.newLinkedHashSet();
-            // add all required facets for updating provider
-            for (Facet requirement : requiredFacets(provider)) {
-                determineProviderChainFor(requirement.value(), result, scalable);
-                List<FacetProvider> requirementChain = result.get(requirement.value());
-                if (requirementChain != null) {
-                    localOrderedProviders.addAll(result.get(requirement.value()));
-                } else {
-                    return;
-                }
-            }
-            // the provider updates this and other facets
-            // just add producers for the other facets
-            for (Facet updated : updatedFacets(provider)) {
-                for (FacetProvider fp : providersList) {
-                    // only add @Produces providers to avoid infinite recursion
-                    if (producesFacet(fp, updated.value())) {
-                        if (!scalable || fp instanceof ScalableFacetProvider) {
-                            localOrderedProviders.add(fp);
-                        } else {
-                            return;
-                        }
+        Stream.of(updatedFacets(provider), requiredFacets(provider)).flatMap(Arrays::stream).forEachOrdered(r -> {
+            if (r.value() != providedFacet) {
```

Comment:
Is that one big `if` around the rest, with no `else`? If so, I'd move this out of the `forEach` and use a `.filter(r -> r.value() == providedFacet)` for that.